framework/aster-frame/src/sync/wait.rs

```diff
@@ -100,80 +103,134 @@ impl WaitQueue {
 }
 
 impl Waiter {
-    pub fn new() -> Self {
-        Waiter {
+    /// Creates a waiter and its associated [`Waker`].
+    pub fn new() -> (Self, Arc<Waker>) {
+        let waker = Arc::new(Waker {
+            has_waken: AtomicBool::new(false),
             task: current_task().unwrap(),
-        }
+        });
+        let waiter = Self {
+            waker: waker.clone(),
+        };
+        (waiter, waker)
     }
 
-    /// Wait until being woken up
+    /// Puts the current thread to sleep until it is woken up by calling [`Waker::wake_up`].
     pub fn wait(&self) {
-        debug_assert_eq!(
-            self.task.inner_exclusive_access().task_status,
-            TaskStatus::Runnable
-        );
-        self.task.inner_exclusive_access().task_status = TaskStatus::Sleeping;
-        while self.task.inner_exclusive_access().task_status == TaskStatus::Sleeping {
-            schedule();
-        }
+        self.waker.do_wait();
+    }
+}
+
+impl Drop for Waiter {
+    fn drop(&mut self) {
+        // When dropping the waiter, we need to close the waker to ensure that if someone wants to
+        // wake up the waiter afterwards, they will perform a no-op.
+        self.waker.close();
     }
+}
 
-    /// Wake up a waiting task.
-    /// If the task is waiting before being woken, return true;
-    /// Otherwise return false.
+impl Waker {
+    /// Wakes up the associated [`Waiter`].
+    ///
+    /// This method returns true if the thread to which the waiter belongs has been woke up from
+    /// sleep.
     pub fn wake_up(&self) -> bool {
+        if self.has_waken.swap(true, Ordering::AcqRel) {
+            return false;
+        }
+
         let mut task = self.task.inner_exclusive_access();
-        if task.task_status == TaskStatus::Sleeping {
-            task.task_status = TaskStatus::Runnable;
+        match task.task_status {
+            TaskStatus::Sleepy => {
+                task.task_status = TaskStatus::Runnable;
+                true
+            }
+            TaskStatus::Sleeping => {
+                task.task_status = TaskStatus::Runnable;
 
-            // Avoid holding lock when doing `add_task`
-            drop(task);
+                // Avoid holding the lock when doing `add_task`
+                drop(task);
+                add_task(self.task.clone());
 
-            add_task(self.task.clone());
+                true
+            }
+            _ => {
+                // Conservatively return `false` to avoid potential race conditions
+                false
+            }
+        }
+    }
 
-            true
-        } else {
-            false
+    fn do_wait(&self) {
+        while !self.has_waken.load(Ordering::Acquire) {
+            let mut task = self.task.inner_exclusive_access();
+            // After holding the lock, check again to avoid races
+            if self.has_waken.load(Ordering::Acquire) {
+                break;
+            }
+            task.task_status = TaskStatus::Sleepy;
```

Comment:
```rust
            task.task_status = TaskStatus::Sleepy;
            drop(task);

            schedule();
```

If the task is woke up after `task.task_status = TaskStatus::Sleepy;` but before `schedule()`, this task should not be added to the task queue. Otherwise, `schedule()` will add the task to the task queue again. That's the problem.

I think adding another task state is the most straightforward way to solve this problem. Do you have other suggestions?

Reply:
Yes, we do need a state for tasks starting waiting but not being scheduled.

Referred to [ThreadStatus in RedLeaf](https://github.com/mars-research/redleaf/blob/7194295d1968c8013ae6b3d104a9192f03516449/lib/core/interfaces/syscalls/src/lib.rs#L104), they also introduce a seperate `Waiting` status. 

How do you like renaming the status as `Waiting`? It seems this status should only used in `WaitQueue`.

Reply:
> This status may be like Sleepy before scheduled, and like Sleeping after scheduled.

I don't understand your suggestion. Without adding a new task state, currently `Sleeping` plays the exact role. Currently, `Sleeping` is used only in `WaitQueue`. Before this PR, it's just like the `Waiting` state that you propose. However, this single state causes the problem that I've described above.

Reply:
> currently Sleeping plays the exact role. Currently, Sleeping is used only in WaitQueue

Thanks for your explanation. I have no further questions regarding this matter.

Reply:
Sorry, but I want to confirmed again, that a running task should always not be in queue of `Scheduler`, right? Since we always dequeue the task and run the task. So the original problem does not actually exist, right?

Reply:
> a running task should always not be in queue of `Scheduler`

Yes.

> So the original problem does not actually exist.

```rust
            task.task_status = TaskStatus::Sleepy;
            drop(task);

            schedule();
```

Suppose that the waiter is woken up between `task.task_status = TaskStatus::Sleepy;` and `schedule()`. It _must not_ be queued because, as you said:

> a running task should always not be in queue of `Scheduler`

That's the point of having a two states (`Sleepy`/`Sleeping` state) to denote whether the task is in the foreground or background.

Reply:
To clarify, "the waiter is woken up" means that the wake event happens in a hardware IRQ or in another CPU core. I think this is the point I didn't explain very clearly and may confuse you.

Reply:
Got it! Thanks for your explanation!